Enhances get snapshots API to allow retrieving repository index only

abeyad opened the pull request:
Currently, the get snapshots API (e.g. /_snapshot/{repositoryName}/_all)
provides information about snapshots in the repository, including the
snapshot state, number of shards snapshotted, failures, etc.  In order
to provide information about each snapshot in the repository, the call
must read the snapshot metadata blob (`snap-{snapshot_uuid}.dat`) for
every snapshot.  In cloud-based repositories, this can be expensive,
both from a cost and performance perspective.  Sometimes, all the user
wants is to retrieve all the names/uuids of each snapshot, and the
indices that went into each snapshot, without any of the other status
information about the snapshot.  This minimal information can be
retrieved from the repository index blob (`index-N`) without needing to
read each snapshot metadata blob.

This commit enhances the get snapshots API with an optional `verbose`
parameter.  If `verbose` is set to false on the request, then the get
snapshots API will only retrieve the minimal information about each
snapshot (the name, uuid, and indices in the snapshot), and only read
this information from the repository index blob, thereby giving users
the option to retrieve the snapshots in a repository in a more
cost-effective and efficient manner.

Closes #24288

StefanSchmidtOz commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -133,4 +155,32 @@ private boolean isAllSnapshots(String[] snapshots) {
+                                                        final List<SnapshotInfo> currentSnapshots) {
+        List<SnapshotInfo> snapshotInfos = new ArrayList<>();
+        for (SnapshotInfo snapshotInfo : currentSnapshots) {
+            if (toResolve.contains(snapshotInfo.snapshotId())) {
```
Idea:
```
if (toResolve.remove(snapshotInfo.snapshotId())) {
        snapshotInfos.add(snapshotInfo.basic());
}
```

abeyad replied:
thank you @StefanSchmidtOz, i made the suggested change

abeyad commented:
@imotov I looked at adding the state to the index blob and I believe this will be relatively straight forward, and I agree with you its only a little bit of added information to the index blob, but the benefits are great as state is the most important thing one would care about for a given snapshot.  I will work on adding this and push commits, and will add the tests you suggested as well.

abeyad commented:
@imotov I pushed https://github.com/elastic/elasticsearch/pull/24477/commits/baa20555b46e79de7851a951803fbba15bdc731d that addresses your feedback and enables storing

imotov commented:
Looks good! Any chance we can add reading non-verbose listings to one of snapshot/restore bwc tests?

abeyad commented:
@imotov the different repository index formats are implicitly executed in the backwards-5.0 and rolling upgrade tests - since we have different versioned nodes, with some of them on the old format, sometimes we get nodes that read/write the old format and sometimes we get nodes that read/write the new format.  In fact, running those tests helped me find a bug in reading the old format from a new node.  

I can add a test which just writes the old x-content and reads it in the newer version, and vice versa (a unit test where I'd be copying/pasting the old code in the test to test old serialization/deserialization with the new).

abeyad commented:
retest this please

abeyad commented:
@imotov i added the old snapshot bwc tests in 9ca8c06e635d3e246c70a288e43a0525f37695cf

imotov commented:
This is going only to 6.0, right? Should we remove 5.5 label?

abeyad commented:
@imotov I thought we could backport it too 5.5 too, do you feel its too risky?

abeyad commented:
retest this please

imotov commented:
@abeyad I don't think it's too risky. I just noticed this
```
VERBOSE_INTRODUCED = Version.V_6_0_0_alpha1_UNRELEASED;
```
but I guess you did it temporary to make sure BWC tests pass.

abeyad commented:
> VERBOSE_INTRODUCED = Version.V_6_0_0_alpha1_UNRELEASED;

Yes exactly, once I backport the change to 5.5, I will change the constant as well on both 5.5 and master.

abeyad commented:
5.x commit: 8fc3f1c1c7bf602fd422052ba29b363d720b9c2a

abeyad commented:
thanks for the review @imotov
